File: macroslib/tests/expectations/cpp_qdate_typemap.cpp
"static QDate f() noexcept;";
"int64_t Foo_f();";

r#"template<bool OWN_DATA>
    inline QDate FooWrapper<OWN_DATA>::f() noexcept
    {

        int64_t ret = Foo_f();
        return QDateTime::fromMSecsSinceEpoch(ret, Qt::UTC, 0).date();
    }"#;

"static std::optional<QDate> f2() noexcept;";
"struct CRustOptioni64 Foo_f2();";
r#"template<bool OWN_DATA>
    inline std::optional<QDate> FooWrapper<OWN_DATA>::f2() noexcept
    {

        struct CRustOptioni64 ret = Foo_f2();
        return (ret.is_some != 0) ? std::optional<QDate>(QDateTime::fromMSecsSinceEpoch(ret.val.data, Qt::UTC, 0).date()) : std::optional<QDate>();
    }"#;

File: macroslib/tests/expectations/cpp_qdate_typemap.rs
foreign_typemap!(
    ($p:r_type) Date<Utc> => i64 {
        $out = $p.add_hms(0, 0, 0).timestamp_millis();
    };
    ($p:f_type) => "QDate"
        "QDateTime::fromMSecsSinceEpoch($p, Qt::UTC, 0).date()";
);

foreigner_class!(class Foo {
    static_method f() -> Date<Utc>;
    static_method f2() -> Option<Date<Utc>>;
});
